// typed-immutable/lib/test/list.js
(function (factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "./test", "immutable", "../record", "../list", "../typed"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("./test"), require("immutable"), require("../record"), require("../list"), require("../typed"));
  }
})(function (exports, _test, _immutable, _record, _list, _typed) {
  "use strict";

  var _createClass = (function () { function defineProperties(target, props) { for (var key in props) { var prop = props[key]; prop.configurable = true; if (prop.value) prop.writable = true; } Object.defineProperties(target, props); } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

  var _inherits = function (subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) subClass.__proto__ = superClass; };

  var _classCallCheck = function (instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } };

  var test = _test["default"];
  var Immutable = _immutable;
  var Record = _record.Record;
  var List = _list.List;
  var Typed = _typed.Typed;
  var Union = _typed.Union;
  var Maybe = _typed.Maybe;

  var NumberList = List(Number);
  var StringList = List(String);
  var Point = Record({ x: Number(0),
    y: Number(0) }, "Point");

  var Points = List(Point, "Points");

  var isUpperCase = function (x) {
    return x.toUpperCase() === x;
  };
  var upperCase = function (x) {
    return x.toUpperCase();
  };
  var inc = function (x) {
    return x + 1;
  };
  var isEvent = function (x) {
    return x % 2 === 0;
  };
  var sum = function (x, y) {
    return x + y;
  };
  var concat = function (xs, ys) {
    return xs.concat(ys);
  };

  test("typed list creation", function (assert) {

    assert.throws(function (_) {
      return List();
    }, /Typed.List must be passed a type descriptor/);

    assert.throws(function (_) {
      return List({});
    }, /Typed.List was passed an invalid type descriptor:/);
  });

  test("number list", function (assert) {
    var ns1 = NumberList();
    assert.ok(ns1 instanceof Immutable.List);
    assert.ok(ns1 instanceof List);
    assert.ok(ns1 instanceof NumberList);
    assert.equal(ns1.size, 0);

    var ns2 = ns1.push(5);
    assert.ok(ns1 instanceof Immutable.List);
    assert.ok(ns1 instanceof List);
    assert.ok(ns1 instanceof NumberList);
    assert.equal(ns2.size, 1);
    assert.equal(ns2.get(0), 5);
    assert.equal(ns2.first(), 5);
    assert.equal(ns2.last(), 5);
  });

  test("empty record list", function (assert) {
    var v = Points();

    assert.ok(v instanceof Immutable.List);
    assert.ok(v instanceof List);
    assert.ok(v instanceof Points);

    assert.equal(v.size, 0);
  });

  test("make list as function call", function (assert) {
    var v = Points([{ x: 1 }]);

    assert.ok(v instanceof Immutable.List);
    assert.ok(v instanceof List);
    assert.ok(v instanceof Points);

    assert.equal(v.size, 1);

    assert.ok(v.get(0) instanceof Record);
    assert.ok(v.get(0) instanceof Point);
    assert.deepEqual(v.toJSON(), [{ x: 1, y: 0 }]);
  });

  test("make list of records", function (assert) {
    var v = Points.of({ x: 10 }, { x: 15 }, { x: 17 });
    assert.ok(v instanceof Immutable.List);
    assert.ok(v instanceof List);
    assert.ok(v instanceof Points);

    assert.equal(v.size, 3);

    assert.ok(v.get(0) instanceof Record);
    assert.ok(v.get(0) instanceof Point);

    assert.ok(v.get(1) instanceof Record);
    assert.ok(v.get(1) instanceof Point);

    assert.ok(v.get(2) instanceof Record);
    assert.ok(v.get(2) instanceof Point);

    assert.deepEqual(v.toJSON(), [{ x: 10, y: 0 }, { x: 15, y: 0 }, { x: 17, y: 0 }]);
  });

  test("make list with new", function (assert) {
    var v = new Points([{ x: 3 }]);

    assert.ok(v instanceof Immutable.List);
    assert.ok(v instanceof List);
    assert.ok(v instanceof Points);

    assert.equal(v.size, 1);

    assert.ok(v.get(0) instanceof Record);
    assert.ok(v.get(0) instanceof Point);
    assert.deepEqual(v.toJSON(), [{ x: 3, y: 0 }]);
  });

  test("toString on typed list", function (assert) {
    var points = Points.of({ x: 10 }, { y: 2 });
    var numbers = NumberList.of(1, 2, 3);
    var strings = StringList.of("hello", "world");

    assert.equal(points.toString(), "Points([ Point({ \"x\": 10, \"y\": 0 }), Point({ \"x\": 0, \"y\": 2 }) ])");

    assert.equal(numbers.toString(), "Typed.List(Number)([ 1, 2, 3 ])");

    assert.equal(strings.toString(), "Typed.List(String)([ \"hello\", \"world\" ])");
  });

  test("create list from entries", function (assert) {
    var ns1 = NumberList.of(1, 2, 3, 4);
    assert.equal(ns1.toString(), "Typed.List(Number)([ 1, 2, 3, 4 ])");
    assert.equal(ns1[Typed.typeName](), "Typed.List(Number)");

    assert.deepEqual(ns1.toJSON(), [1, 2, 3, 4]);
  });

  test("converts sequences to list", function (assert) {
    var seq = Immutable.Seq([{ x: 1 }, { x: 2 }]);
    var v = Points(seq);

    assert.ok(v instanceof Immutable.List);
    assert.ok(v instanceof List);
    assert.ok(v instanceof Points);

    assert.equal(v.size, 2);

    assert.ok(v.get(0) instanceof Record);
    assert.ok(v.get(0) instanceof Point);
    assert.ok(v.get(1) instanceof Record);
    assert.ok(v.get(1) instanceof Point);

    assert.deepEqual(v.toJSON(), [{ x: 1, y: 0 }, { x: 2, y: 0 }]);
  });

  test("can be subclassed", function (assert) {
    var Graph = (function (_Points) {
      function Graph() {
        _classCallCheck(this, Graph);

        if (_Points != null) {
          _Points.apply(this, arguments);
        }
      }

      _inherits(Graph, _Points);

      _createClass(Graph, {
        foo: {
          value: function foo() {
            var first = this.first();
            var last = this.last();
            return last.x - first.x;
          }
        }
      });

      return Graph;
    })(Points);

    var v1 = new Graph([{ y: 3 }, { x: 7 }, { x: 9, y: 4 }]);

    assert.ok(v1 instanceof Immutable.List);
    assert.ok(v1 instanceof List);
    assert.ok(v1 instanceof Points);
    assert.ok(v1 instanceof Graph);

    assert.equal(v1.foo(), 9);
    assert.deepEqual(v1.toJSON(), [{ x: 0, y: 3 }, { x: 7, y: 0 }, { x: 9, y: 4 }]);

    var v2 = v1.set(0, { x: 2, y: 4 });

    assert.ok(v2 instanceof Immutable.List);
    assert.ok(v2 instanceof List);
    assert.ok(v2 instanceof Points);
    assert.ok(v2 instanceof Graph);

    assert.equal(v2.foo(), 7);
    assert.deepEqual(v2.toJSON(), [{ x: 2, y: 4 }, { x: 7, y: 0 }, { x: 9, y: 4 }]);
  });

  test("short-circuits if already a list", function (assert) {
    var v1 = Points.of({ x: 2, y: 4 }, { x: 8, y: 3 });

    assert.equal(v1, Points(v1));

    assert.equal(v1, new Points(v1));

    var OtherPoints = List(Point);

    assert.ok(OtherPoints(v1) instanceof OtherPoints);
    assert.notOk(OtherPoints(v1) instanceof Points);
    assert.notEqual(v1, OtherPoints(v1));
    assert.ok(v1.equals(OtherPoints(v1)));

    assert.ok(new OtherPoints(v1) instanceof OtherPoints);
    assert.notOk(new OtherPoints(v1) instanceof Points);
    assert.notEqual(v1, new OtherPoints(v1));
    assert.ok(v1.equals(new OtherPoints(v1)));

    var SubPoints = (function (_Points) {
      function SubPoints() {
        _classCallCheck(this, SubPoints);

        if (_Points != null) {
          _Points.apply(this, arguments);
        }
      }

      _inherits(SubPoints, _Points);

      _createClass(SubPoints, {
        head: {
          value: function head() {
            return this.first();
          }
        }
      });

      return SubPoints;
    })(Points);

    assert.notEqual(v1, new SubPoints(v1));
    assert.ok(v1.equals(new SubPoints(v1)));

    assert.equal(new SubPoints(v1).head(), v1.first());
  });

  test("can be cleared", function (assert) {
    var v1 = Points.of({ x: 1 }, { x: 2 }, { x: 3 });
    var v2 = v1.clear();

    assert.ok(v1 instanceof Points);
    assert.ok(v2 instanceof Points);

    assert.equal(v1.size, 3);
    assert.equal(v2.size, 0);

    assert.deepEqual(v1.toJSON(), [{ x: 1, y: 0 }, { x: 2, y: 0 }, { x: 3, y: 0 }]);

    assert.deepEqual(v2.toJSON(), []);

    assert.equal(v2.first(), void 0);
  });

  test("can construct records", function (assert) {
    var v1 = Points();
    var v2 = v1.push({ x: 1 });
    var v3 = v2.push({ y: 2 });
    var v4 = v3.push({ x: 3, y: 3 });
    var v5 = v4.push(void 0);

    assert.ok(v1 instanceof Points);
    assert.ok(v2 instanceof Points);
    assert.ok(v3 instanceof Points);
    assert.ok(v4 instanceof Points);
    assert.ok(v5 instanceof Points);

    assert.equal(v1.size, 0);
    assert.equal(v2.size, 1);
    assert.equal(v3.size, 2);
    assert.equal(v4.size, 3);
    assert.equal(v5.size, 4);

    assert.deepEqual(v1.toJSON(), []);
    assert.deepEqual(v2.toJSON(), [{ x: 1, y: 0 }]);
    assert.deepEqual(v3.toJSON(), [{ x: 1, y: 0 }, { x: 0, y: 2 }]);
    assert.deepEqual(v4.toJSON(), [{ x: 1, y: 0 }, { x: 0, y: 2 }, { x: 3, y: 3 }]);
    assert.deepEqual(v5.toJSON(), [{ x: 1, y: 0 }, { x: 0, y: 2 }, { x: 3, y: 3 }, { x: 0, y: 0 }]);
  });

  test("can update sub-records", function (assert) {
    var v1 = Points.of({ x: 4 }, { y: 4 });
    var v2 = v1.setIn([0, "y"], 5);
    var v3 = v2.set(2, void 0);
    var v4 = v3.setIn([1, "y"], void 0);

    assert.ok(v1 instanceof Points);
    assert.ok(v2 instanceof Points);
    assert.ok(v3 instanceof Points);
    assert.ok(v4 instanceof Points);

    assert.equal(v1.size, 2);
    assert.equal(v2.size, 2);
    assert.equal(v3.size, 3);
    assert.equal(v4.size, 3);

    assert.deepEqual(v1.toJSON(), [{ x: 4, y: 0 }, { x: 0, y: 4 }]);

    assert.deepEqual(v2.toJSON(), [{ x: 4, y: 5 }, { x: 0, y: 4 }]);

    assert.deepEqual(v3.toJSON(), [{ x: 4, y: 5 }, { x: 0, y: 4 }, { x: 0, y: 0 }]);

    assert.deepEqual(v4.toJSON(), [{ x: 4, y: 5 }, { x: 0, y: 0 }, { x: 0, y: 0 }]);
  });

  test("serialize & parse", function (assert) {
    var ns1 = NumberList.of(1, 2, 3, 4);

    assert.ok(NumberList(ns1.toJSON()).equals(ns1), "parsing serialized typed list");

    assert.ok(ns1.constructor(ns1.toJSON()).equals(ns1), "parsing with constructor");
  });

  test("serialize & parse nested", function (assert) {
    var v1 = Points.of({ x: 1 }, { x: 2 }, { y: 3 });

    assert.ok(Points(v1.toJSON()).equals(v1));
    assert.ok(v1.constructor(v1.toJSON()).equals(v1));
    assert.ok(v1.equals(new Points(v1.toJSON())));

    assert.ok(Points(v1.toJSON()).get(0) instanceof Point);
  });

  test("construct with array", function (assert) {
    var ns1 = NumberList([1, 2, 3, 4, 5]);

    assert.ok(ns1 instanceof NumberList);
    assert.ok(ns1.size, 5);
    assert.equal(ns1.get(0), 1);
    assert.equal(ns1.get(1), 2);
    assert.equal(ns1.get(2), 3);
    assert.equal(ns1.get(3), 4);
    assert.equal(ns1.get(4), 5);
  });

  test("construct with indexed seq", function (assert) {
    var seq = Immutable.Seq([1, 2, 3]);
    var ns1 = NumberList(seq);

    assert.ok(ns1 instanceof NumberList);
    assert.ok(ns1.size, 3);
    assert.equal(ns1.get(0), 1);
    assert.equal(ns1.get(1), 2);
    assert.equal(ns1.get(2), 3);
  });

  test("does not construct form a scalar", function (assert) {
    assert.throws(function (_) {
      return NumberList(3);
    }, /Expected Array or iterable object of values/);
  });

  test("can not construct with invalid data", function (assert) {
    var Point = Record({ x: Number, y: Number }, "Point");
    var Points = List(Point, "Points");

    assert.throws(function (_) {
      return Points.of({ x: 1, y: 0 }, { y: 2, x: 2 }, { x: 3 });
    }, /"undefined" is not a number/);
  });

  test("set and get a value", function (assert) {
    var ns = NumberList();
    var ns2 = ns.set(0, 7);

    assert.equal(ns.size, 0);
    assert.equal(ns.count(), 0);
    assert.equal(ns.get(0), void 0);

    assert.equal(ns2.size, 1);
    assert.equal(ns2.count(), 1);
    assert.equal(ns2.get(0), 7);
  });

  test("set and get records", function (assert) {
    var v1 = Points();
    var v2 = v1.set(0, { x: 7 });

    assert.equal(v1.size, 0);
    assert.equal(v1.count(), 0);
    assert.equal(v1.get(0), void 0);

    assert.equal(v2.size, 1);
    assert.equal(v2.count(), 1);
    assert.ok(v2.get(0) instanceof Point);
    assert.ok(v2.get(0).toJSON(), { x: 7, y: 0 });
  });

  test("can not set invalid value", function (assert) {
    var ns = NumberList();

    assert.throws(function (_) {
      return ns.set(0, "foo");
    }, /"foo" is not a number/);

    assert.equal(ns.size, 0);
  });

  test("can not set invalid structure", function (assert) {
    var v = Points();

    assert.throws(function (_) {
      return v.set(0, 5);
    }, /Invalid data structure/);
  });

  test("can not set undeclared fields", function (assert) {
    var v = Points.of({ x: 9 });

    assert.throws(function (_) {
      return v.setIn([0, "d"], 4);
    }, /Cannot set unknown field "d"/);
  });

  test("counts from the end of the list on negative index", function (assert) {
    var ns = NumberList.of(1, 2, 3, 4, 5, 6, 7);
    assert.equal(ns.get(-1), 7);
    assert.equal(ns.get(-5), 3);
    assert.equal(ns.get(-9), void 0);
    assert.equal(ns.get(-999, 1000), 1000);
  });

  test("coerces numeric-string keys", function (assert) {
    // Of course, TypeScript protects us from this, so cast to "any" to test.
    var ns = NumberList.of(1, 2, 3, 4, 5, 6);

    assert.equal(ns.get("1"), 2);
    assert.equal(ns.get("-1"), 6);
    assert.equal(ns.set("3", 10).get("-3"), 10);
  });

  test("setting creates a new instance", function (assert) {
    var v1 = NumberList.of(1);
    var v2 = v1.set(0, 15);

    assert.equal(v1.get(0), 1);
    assert.equal(v2.get(0), 15);

    assert.ok(v1 instanceof NumberList);
    assert.ok(v2 instanceof NumberList);
  });

  test("size includes the highest index", function (assert) {
    var v0 = NumberList();
    var v1 = v0.set(0, 1);
    var v2 = v1.set(1, 2);
    var v3 = v2.set(2, 3);

    assert.equal(v0.size, 0);
    assert.equal(v1.size, 1);
    assert.equal(v2.size, 2);
    assert.equal(v3.size, 3);

    assert.ok(v0 instanceof NumberList);
    assert.ok(v1 instanceof NumberList);
    assert.ok(v2 instanceof NumberList);
    assert.ok(v3 instanceof NumberList);
  });

  test("get helpers make for easier to read code", function (assert) {
    var v1 = NumberList.of(1, 2, 3);

    assert.equal(v1.first(), 1);
    assert.equal(v1.get(1), 2);
    assert.equal(v1.last(), 3);
  });

  test("slice helpers make for easier to read code", function (assert) {
    var v0 = NumberList.of(1, 2, 3);
    var v1 = NumberList.of(1, 2);
    var v2 = NumberList.of(1);
    var v3 = NumberList();

    assert.deepEqual(v0.rest().toArray(), [2, 3]);
    assert.ok(v0.rest() instanceof NumberList);
    assert.deepEqual(v0.butLast().toArray(), [1, 2]);
    assert.ok(v0.butLast() instanceof NumberList);

    assert.deepEqual(v1.rest().toArray(), [2]);
    assert.ok(v1.rest() instanceof NumberList);
    assert.deepEqual(v1.butLast().toArray(), [1]);
    assert.ok(v1.butLast() instanceof NumberList);

    assert.deepEqual(v2.rest().toArray(), []);
    assert.ok(v2.rest() instanceof NumberList);
    assert.deepEqual(v2.butLast().toArray(), []);
    assert.ok(v2.butLast() instanceof NumberList);

    assert.deepEqual(v3.rest().toArray(), []);
    assert.ok(v3.rest() instanceof NumberList);
    assert.deepEqual(v3.butLast().toArray(), []);
    assert.ok(v3.butLast() instanceof NumberList);
  });

  test("can set at with in the bonds", function (assert) {
    var v0 = NumberList.of(1, 2, 3);
    var v1 = v0.set(1, 20); // within existing tail
    var v2 = v1.set(3, 30); // at last position

    assert.throws(function (_) {
      return v1.set(4, 4);
    }, /Index "4" is out of bound/);
    assert.throws(function (_) {
      return v2.set(31, 31);
    }, /Index "31" is out of bound/);

    assert.equal(v2.size, v1.size + 1);

    assert.deepEqual(v0.toArray(), [1, 2, 3]);
    assert.deepEqual(v1.toArray(), [1, 20, 3]);
    assert.deepEqual(v2.toArray(), [1, 20, 3, 30]);

    assert.ok(v0 instanceof NumberList);
    assert.ok(v1 instanceof NumberList);
    assert.ok(v2 instanceof NumberList);
  });

  test("can contain a large number of indices", function (assert) {
    var input = Immutable.Range(0, 20000);
    var numbers = NumberList(input);
    var iterations = 0;

    assert.ok(numbers.every(function (value) {
      var result = value === iterations;
      iterations = iterations + 1;
      return result;
    }));
  });

  test("push inserts at highest index", function (assert) {
    var v0 = NumberList.of(1, 2, 3);
    var v1 = v0.push(4, 5, 6);

    assert.ok(v0 instanceof NumberList);
    assert.ok(v1 instanceof NumberList);

    assert.equal(v0.size, 3);
    assert.equal(v1.size, 6);

    assert.deepEqual(v0.toArray(), [1, 2, 3]);
    assert.deepEqual(v1.toArray(), [1, 2, 3, 4, 5, 6]);
  });

  test("pop removes the highest index, decrementing size", function (assert) {
    var v0 = NumberList.of(1, 2, 3);
    var v1 = v0.pop();
    var v2 = v1.push(4);

    assert.equal(v0.last(), 3);
    assert.equal(v0.size, 3);
    assert.deepEqual(v0.toArray(), [1, 2, 3]);

    assert.ok(v1 instanceof NumberList);
    assert.equal(v1.last(), 2);
    assert.equal(v1.size, 2);
    assert.deepEqual(v1.toArray(), [1, 2]);

    assert.ok(v2 instanceof NumberList);
    assert.equal(v2.last(), 4);
    assert.equal(v2.size, 3);
    assert.deepEqual(v2.toArray(), [1, 2, 4]);
  });

  test("pop on empty", function (assert) {
    var v0 = NumberList.of(1);
    var v1 = v0.pop();
    var v2 = v1.pop();
    var v3 = v2.pop();
    var v4 = v3.pop();
    var v5 = v4.pop();

    assert.equal(v0.size, 1);
    assert.deepEqual(v0.toArray(), [1]);

    ![v1, v2, v3, v4, v5].forEach(function (v) {
      assert.ok(v instanceof NumberList);
      assert.equal(v.size, 0);
      assert.deepEqual(v.toArray(), []);
    });
  });

  test("test removes any index", function (assert) {
    var v0 = NumberList.of(1, 2, 3);
    var v1 = v0.remove(2);
    var v2 = v1.remove(0);
    var v3 = v2.remove(9);
    var v4 = v0.remove(3);
    var v5 = v3.push(5);

    assert.ok(v0 instanceof NumberList);
    assert.ok(v1 instanceof NumberList);
    assert.ok(v2 instanceof NumberList);
    assert.ok(v3 instanceof NumberList);
    assert.ok(v4 instanceof NumberList);
    assert.ok(v5 instanceof NumberList);

    assert.equal(v0.size, 3);
    assert.equal(v1.size, 2);
    assert.equal(v2.size, 1);
    assert.equal(v3.size, 1);
    assert.equal(v4.size, 3);
    assert.equal(v5.size, 2);

    assert.deepEqual(v0.toArray(), [1, 2, 3]);
    assert.deepEqual(v1.toArray(), [1, 2]);
    assert.deepEqual(v2.toArray(), [2]);
    assert.deepEqual(v3.toArray(), [2]);
    assert.deepEqual(v4.toArray(), [1, 2, 3]);
    assert.deepEqual(v5.toArray(), [2, 5]);
  });

  test("shift removes from the front", function (assert) {
    var v0 = NumberList.of(1, 2, 3);
    var v1 = v0.shift();

    assert.ok(v0 instanceof NumberList);
    assert.ok(v1 instanceof NumberList);

    assert.deepEqual(v0.toArray(), [1, 2, 3]);
    assert.deepEqual(v1.toArray(), [2, 3]);

    assert.equal(v0.first(), 1);
    assert.equal(v1.first(), 2);

    assert.equal(v0.size, 3);
    assert.equal(v1.size, 2);
  });

  test("unshift insert items in the front", function (assert) {
    var v0 = NumberList.of(1, 2, 3);
    var v1 = v0.unshift(11, 12, 13);

    assert.ok(v0 instanceof NumberList);
    assert.ok(v1 instanceof NumberList);

    assert.deepEqual(v0.toArray(), [1, 2, 3]);
    assert.deepEqual(v1.toArray(), [11, 12, 13, 1, 2, 3]);

    assert.equal(v0.first(), 1);
    assert.equal(v1.first(), 11);

    assert.equal(v0.size, 3);
    assert.equal(v1.size, 6);
  });

  test("finds values using indexOf", function (assert) {
    var v = NumberList.of(1, 2, 3, 2, 1);

    assert.equal(v.indexOf(2), 1);
    assert.equal(v.indexOf(3), 2);
    assert.equal(v.indexOf(4), -1);
  });

  test("finds values using findIndex", function (assert) {
    var v = StringList.of("a", "b", "c", "B", "a");

    assert.equal(v.findIndex(isUpperCase), 3);
    assert.equal(v.findIndex(function (x) {
      return x.length > 1;
    }), -1);
  });

  test("finds values using findEntry", function (assert) {
    var v = StringList.of("a", "b", "c", "B", "a");

    assert.deepEqual(v.findEntry(isUpperCase), [3, "B"]);
    assert.equal(v.findEntry(function (x) {
      return x.length > 1;
    }), void 0);
  });

  test("maps values", function (assert) {
    var v0 = NumberList.of(1, 2, 3);
    var v1 = v0.map(inc);

    assert.ok(v0 instanceof NumberList);
    assert.ok(v1 instanceof NumberList);
    assert.ok(v1 instanceof Immutable.List);

    assert.equal(v0.size, 3);
    assert.equal(v1.size, 3);

    assert.deepEqual(v0.toArray(), [1, 2, 3]);
    assert.deepEqual(v1.toArray(), [2, 3, 4]);
  });

  test("maps records to any", function (assert) {
    var v0 = Points.of({ x: 1 }, { y: 2 }, { x: 3, y: 3 });
    var v1 = v0.map(function (_ref) {
      var x = _ref.x;
      var y = _ref.y;
      return { x: x + 1, y: y * y };
    });

    assert.ok(v0 instanceof Points);
    assert.notOk(v1 instanceof Points);
    assert.ok(v1 instanceof Immutable.List);
    assert.equal(v1[Typed.typeName](), "Typed.List(Any)");

    assert.equal(v0.size, 3);
    assert.equal(v1.size, 3);

    assert.deepEqual(v0.toJSON(), [{ x: 1, y: 0 }, { x: 0, y: 2 }, { x: 3, y: 3 }]);

    assert.deepEqual(v1.toJSON(), [{ x: 2, y: 0 }, { x: 1, y: 4 }, { x: 4, y: 9 }]);
  });

  test("maps records to records", function (assert) {
    var v0 = Points.of({ x: 1 }, { y: 2 }, { x: 3, y: 3 });
    var v1 = v0.map(function (point) {
      return point.update("x", inc).update("y", inc);
    });

    assert.ok(v0 instanceof Points);
    assert.ok(v1 instanceof Points);
    assert.ok(v1 instanceof Immutable.List);

    assert.equal(v0.size, 3);
    assert.equal(v1.size, 3);

    assert.deepEqual(v0.toJSON(), [{ x: 1, y: 0 }, { x: 0, y: 2 }, { x: 3, y: 3 }]);

    assert.deepEqual(v1.toJSON(), [{ x: 2, y: 1 }, { x: 1, y: 3 }, { x: 4, y: 4 }]);
  });

  test("filters values", function (assert) {
    var v0 = NumberList.of(1, 2, 3, 4, 5, 6);
    var v1 = v0.filter(isEvent);

    assert.ok(v0 instanceof NumberList);
    assert.ok(v1 instanceof NumberList);

    assert.equal(v0.size, 6);
    assert.equal(v1.size, 3);

    assert.deepEqual(v0.toArray(), [1, 2, 3, 4, 5, 6]);
    assert.deepEqual(v1.toArray(), [2, 4, 6]);
  });

  test("reduces values", function (assert) {
    var v = NumberList.of(1, 10, 100);

    assert.equal(v.reduce(sum), 111);
    assert.equal(v.reduce(sum, 1000), 1111);

    assert.ok(v instanceof NumberList);
    assert.deepEqual(v.toArray(), [1, 10, 100]);
  });

  test("reduces from the right", function (assert) {
    var v = StringList.of("a", "b", "c");

    assert.equal(v.reduceRight(concat), "cba");
    assert.equal(v.reduceRight(concat, "seeded"), "seededcba");

    assert.ok(v instanceof StringList);
    assert.deepEqual(v.toArray(), ["a", "b", "c"]);
  });

  test("takes and skips values", function (assert) {
    var v0 = NumberList.of(1, 2, 3, 4, 5, 6);
    var v1 = v0.skip(2);
    var v2 = v1.take(2);

    assert.ok(v0 instanceof NumberList);
    assert.ok(v1 instanceof NumberList);
    assert.ok(v2 instanceof NumberList);

    assert.equal(v0.size, 6);
    assert.equal(v1.size, 4);
    assert.equal(v2.size, 2);

    assert.deepEqual(v0.toArray(), [1, 2, 3, 4, 5, 6]);
    assert.deepEqual(v1.toArray(), [3, 4, 5, 6]);
    assert.deepEqual(v2.toArray(), [3, 4]);
  });

  test("efficiently chains array methods", function (assert) {
    var v = NumberList.of(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14);

    assert.equal(v.filter(function (x) {
      return x % 2 == 0;
    }).skip(2).map(function (x) {
      return x * x;
    }).take(3).reduce(function (a, b) {
      return a + b;
    }, 0), 200);

    assert.ok(v instanceof NumberList);
    assert.equal(v.size, 14);
    assert.deepEqual(v.toArray(), [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14]);
  });

  test("convert to map", function (assert) {
    var v = StringList.of("a", "b", "c");
    var m = v.toMap();

    assert.ok(v instanceof StringList);
    assert.equal(v.size, 3);
    assert.deepEqual(v.toArray(), ["a", "b", "c"]);

    assert.equal(m.size, 3);
    assert.equal(m.get(1), "b");
  });

  test("reverses", function (assert) {
    var v0 = StringList.of("a", "b", "c");
    var v1 = v0.reverse();

    assert.ok(v0 instanceof StringList);
    assert.ok(v1 instanceof StringList);

    assert.equal(v0.size, 3);
    assert.equal(v1.size, 3);

    assert.deepEqual(v0.toArray(), ["a", "b", "c"]);
    assert.deepEqual(v1.toArray(), ["c", "b", "a"]);
  });

  test("ensures equality", function (assert) {
    // Make a sufficiently long list.
    var array = Array(100).join("abcdefghijklmnopqrstuvwxyz").split("");

    var v1 = StringList(array);
    var v2 = StringList(array);

    assert.ok(v1 != v2);
    assert.ok(v1.equals(v2));
  });

  test("concat works like Array.prototype.concat", function (assert) {
    var v1 = NumberList.of(1, 2, 3);
    var v2 = v1.concat(4, NumberList.of(5, 6), [7, 8], Immutable.Seq({ a: 9, b: 10 }), Immutable.Set.of(11, 12));

    assert.ok(v1 instanceof NumberList);
    assert.ok(v2 instanceof NumberList);

    assert.equal(v1.size, 3);
    assert.equal(v2.size, 12);

    assert.deepEqual(v1.toArray(), [1, 2, 3]);
    assert.deepEqual(v2.toArray(), [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]);
  });

  test("allows chained mutations", function (assert) {
    var v1 = NumberList();
    var v2 = v1.push(1);
    var v3 = v2.withMutations(function (v) {
      return v.push(2).push(3).push(4);
    });
    var v4 = v3.push(5);

    assert.ok(v1 instanceof NumberList);
    assert.ok(v2 instanceof NumberList);
    assert.ok(v3 instanceof NumberList);
    assert.ok(v4 instanceof NumberList);

    assert.equal(v1.size, 0);
    assert.equal(v2.size, 1);
    assert.equal(v3.size, 4);
    assert.equal(v4.size, 5);

    assert.deepEqual(v1.toArray(), []);
    assert.deepEqual(v2.toArray(), [1]);
    assert.deepEqual(v3.toArray(), [1, 2, 3, 4]);
    assert.deepEqual(v4.toArray(), [1, 2, 3, 4, 5]);
  });

  test("allows chained mutations using alternative API", function (assert) {
    var v1 = NumberList();
    var v2 = v1.push(1);
    var v3 = v2.asMutable().push(2).push(3).push(4).asImmutable();
    var v4 = v3.push(5);

    assert.ok(v1 instanceof NumberList);
    assert.ok(v2 instanceof NumberList);
    assert.ok(v3 instanceof NumberList);
    assert.ok(v4 instanceof NumberList);

    assert.equal(v1.size, 0);
    assert.equal(v2.size, 1);
    assert.equal(v3.size, 4);
    assert.equal(v4.size, 5);

    assert.deepEqual(v1.toArray(), []);
    assert.deepEqual(v2.toArray(), [1]);
    assert.deepEqual(v3.toArray(), [1, 2, 3, 4]);
    assert.deepEqual(v4.toArray(), [1, 2, 3, 4, 5]);
  });

  test("allows size to be set", function (assert) {
    var input = Immutable.Range(0, 2000);
    var v1 = NumberList(input);
    var v2 = v1.setSize(1000);
    assert.throws(function (_) {
      return v2.setSize(1500);
    }, /setSize may only downsize/);
    var v3 = v2.setSize(1000);

    assert.ok(v1 instanceof NumberList);
    assert.ok(v2 instanceof NumberList);
    assert.ok(v3 instanceof NumberList);

    assert.equal(v1.size, 2000);
    assert.equal(v2.size, 1000);
    assert.equal(v3.size, 1000);

    assert.equal(v1.get(900), 900);
    assert.equal(v1.get(1300), 1300);
    assert.equal(v1.get(1800), 1800);

    assert.equal(v2.get(900), 900);
    assert.equal(v2.get(1300), void 0);
    assert.equal(v2.get(1800), void 0);

    assert.equal(v3.get(900), 900);
    assert.equal(v3.get(1300), void 0);
    assert.equal(v3.get(1800), void 0);

    assert.ok(v2.equals(v3));
  });

  test("can be efficiently sliced", function (assert) {
    var input = Immutable.Range(0, 2000);
    var v1 = NumberList(input);
    var v2 = v1.slice(100, -100);

    assert.ok(v1 instanceof NumberList);
    assert.ok(v2 instanceof NumberList);

    assert.equal(v1.size, 2000);
    assert.equal(v2.size, 1800);

    assert.equal(v2.first(), 100);
    assert.equal(v2.rest().size, 1799);
    assert.equal(v2.last(), 1899);
    assert.equal(v2.butLast().size, 1799);
  });
});
